### PR TITLE
179: add ecoregional density observations

### PR DESCRIPTION
**Pull request by thanrobinson**

### CODE REVIEW

**Comment by thanrobinson** [comment]
[Density Tables](https://trello.com/c/DdtO5OTi/179-density-tables)

File: src/task_scl_eff_pot_hab.py [hunk 1]
```diff
@@ -131,12 +131,33 @@ def calc(self):
...
+            med_density_biome = ee.Number(ft.get('MED_DENSITY_ECO'))
+            med_density_eco_biome = ee.Number(ft.get('MED_DENSITY_ECO'))
+            density_val = ee.Algorithms.If(med_density_eco.gt(0),
+                med_density_eco,
+                ee.Algorithms.If(med_density_biome.gt(0),
+                    med_density_biome,
+                    ee.Algorithms.If(med_density_eco_biome.gt(0),
+                        med_density_eco_biome,
+                        1)))
+            min_core_size = ee.Number(500).divide(density_val).int()
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: Should abstract hardcoded values into class properties

File: src/task_scl_eff_pot_hab.py [hunk 2]
```diff
@@ -131,12 +131,33 @@ def calc(self):
...
+        ecoregion_image = self.ecoregions.reduceToImage(
+            properties=['ECO_ID'],
+            reducer=ee.Reducer.mode()
+        )
+
+        ecoregion_id = density.aggregate_array('ECO_ID')
+        core_size = density.aggregate_array('min_core_size')
+
+        min_patch_size = ecoregion_image.remap(
+            ecoregion_id, core_size).int().clamp(30, 625).updateMask(hii_ecoregion)
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: Should abstract hardcoded values into class properties (next line too)

File: src/task_scl_eff_pot_hab.py [hunk 3]
```diff
@@ -181,7 +202,7 @@ def calc(self):
             ],
         )
 
-        scl_eff_pot_hab = "{}/geographies/Sumatra/scl_poly/{}/scl_eff_pot_hab".format(
+        scl_eff_pot_hab = "{}/geographies/Sumatra/scl_poly/{}/scl_eff_pot_hab_test".format(
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: remove temporary `_test`

File: src/task_scl_eff_pot_hab.py [hunk 4]
```diff
@@ -131,12 +131,33 @@ def calc(self):
...
 
-        min_patch_size = self.density.remap(
-            self.remap_lists["density"], self.remap_lists["min_patch_size"]
-        ).updateMask(hii_ecoregion)
-        min_stepping_stone_size = self.density.remap(
-            self.remap_lists["density"], self.remap_lists["min_stepping_stone_size"]
-        ).updateMask(hii_ecoregion)
+        def density_to_patch_size(ft):
+            med_density_eco = ee.Number(ft.get('MED_DENSITY_ECO'))
+            med_density_biome = ee.Number(ft.get('MED_DENSITY_ECO'))
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: I'm not understanding here... shouldn't this and the next line be pointing to different values?

File: src/task_scl_eff_pot_hab.py [hunk 5]
```diff
@@ -21,7 +20,7 @@ class SCLPolygons(SCLTask):
         "probability": {
             "ee_type": SCLTask.IMAGECOLLECTION,
             "ee_path": "projects/SCL/v1/Panthera_tigris/geographies/Sumatra/hab/probability",
-            "maxage": 1, 
+            "maxage": 1,
         },
         "historic_range": {
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py:26]: after bringing in the new `task_base` this should point to `projects/SCL/v1/Panthera_tigris/aoi` and then we rasterize as part of task

File: src/task_scl_eff_pot_hab.py [hunk 6]
```diff
@@ -21,7 +20,7 @@ class SCLPolygons(SCLTask):
         "probability": {
             "ee_type": SCLTask.IMAGECOLLECTION,
             "ee_path": "projects/SCL/v1/Panthera_tigris/geographies/Sumatra/hab/probability",
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py:23]: this will change after moving to new `task_base` paths as discussed

File: src/task_scl_eff_pot_hab.py [hunk 7]
```diff
@@ -48,9 +47,14 @@ class SCLPolygons(SCLTask):
             "ee_path": "projects/SCL/v1/Panthera_tigris/source/Inputs_2006/eco_cntry",
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py:48]: this should point to `USDOS/LSIB/2013` and then we rasterize as part of task
For this and for ecoregions, we should both try and think of a way to have tasks share common definitions

**Review by saanobhaai — CHANGES_REQUESTED** [review]
The first path comments are fyi for next task only

File: src/task_scl_eff_pot_hab.py (hunk 4 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: Yes. Fixed.

File: src/task_scl_eff_pot_hab.py (hunk 1 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: will update this

File: src/task_scl_eff_pot_hab.py (hunk 3 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: noted

**Pull request merged**